_posts/2017-04-30-React-Native-for-iOS-devs.markdown

```diff
@@ -0,0 +1,671 @@
+
+That is the core idea of React Native. Bridge the React component tree to native primitives. React Native runs on a lot of platforms:
+
+-   Officially: iOS, Android, tvOS & VR.
+-   Unofficially: macOS, Windows & Ubuntu.
+
+Each of these platforms will have their own way of showing some text e.g.
+
+-   `RCTText` for iOS and tvOS - [which uses NSTextStorage, and drawRect][1]
+-   `Textfield` for Android - which uses [Canvas and a DrawCommand][2]
+-   `Three.js view primitive` for VR - which uses [BitmapFontGeometry][3] + [Shaders][4]
+-   `RCTText` for macOS [which also uses NSTextStorage, and drawRect][5]
+-   `ReactTextShadowNode` for Windows - which uses a [RichTextBlock][6]
+-   `QQuickItem` for Ubuntu - Which uses [QString to render][7]
+
+But when working at React-level, you would use the component `Text`. This means you work at a "React in JS" level, and rely on the primitives provided by each implementation of React Native.
+
+For iOS, this works by using a JavaScript runtime (running via JavaScriptCore in your app) which sends [messages across a bridge][nick-msg] that handles the native `UIView` hierarchy. Most of the messaging work is handled inside the `RCTUIViewManager` which receives calls like `createView:viewName:rootTag:props:`, `setChildren:reactTags:`, `updateView:viewName:props:` and `createAnimatedNode:config`.
+
+This bridging is how you get a lot of the positive aspects of working with the JavaScript tooling ecosystem. The JavaScript used by React can be updated independent of the app, but so long as it is working with the same native bridge version. This bridging technique is how React can safely have a reliable version of [Injection for Xcode][]. It re-evaluates your JavaScript code, and that triggers a new set of messages to the native side.
+
+Like any cross-platform abstraction, React Native can be leaky. To write a cross-platform app that purely lives inside JS Runtime, you have to write React-only code. React and React Native doesn't have ways to handle primitives like `UINavigationController` - they want your entire app to be represented as a series of components that can be mapped across many platforms. 
+
+This isn't optimal when you're coming in from the native world - where you're used to building platform-specific experiences, and are genuinely excited at the prospect of platform-specific APIs. Generally you can look for other teams who have felt the same and are willing to write native-bridged code that's specific to iOS. Shout-out to [Wix][] and [AirBnB][] who are doing great work in this space.
+
+Is this a critical problem against React Native? I don't think so, we've added native abstractions where it was the right decision and we've used JavaScript when it was the right decision. For example, our `Image` component is a bridged native component that uses `SDWebImage` under the hood so that we can share an image cache with the rest of the app.
```

Comment:
I think it would be great to make a bigger point about this. It's not about writing entire apps just in JS, but using what makes sense for any given problem. I think there is too much drama around going all-in on JS, which is besides the point. Speed up development of all trivial UI, but then go back into native when you have to.